sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,765 @@
+     %A.i32 = OpSpecConstant %int.type 0                           ; 1st member with default value
+   %A.float = OpSpecConstant %float.type 0.0                       ; 2nd member with default value
+    %struct = OpSpecConstantComposite %struct.type %A.i32 %A.float ; Composite doens't need IDs or default value
+         %1 = OpTypeFunction %int
+
+       %get = OpFunction %int None %1
+         %2 = OpLabel
+              OpReturnValue %i32
+              OpFunctionEnd
+         %1 = OpTypeFunction %struct.type
+
+      %get2 = OpFunction %struct.type None %struct
+         %2 = OpLabel
+              OpReturnValue %struct
+              OpFunctionEnd
+```
+
+### Emulation of specialization constants
+
+Emulation of specialization constants is performed by converting them into
+kernel arguments.
+
+Overall idea is that DPC++ runtimes packs all specialization constants into a
+single buffer, which is passed as an extra implicit kernel argument. Then the
+compiler instead of lowering `__sycl_get*2020SpecConstantValue` intrinsics into
+SPIR-V friendly IR replaces it with extracting an element from that buffer.
+
+"All" specialization constants here means complete list of specialization
+constants encountered in an application or a shared library which is being
+compiled: that list is computed by `sycl-post-link` tool and communicated to
+the runtime through device image properties like it is described in "Support for
+native specialization constants" section.
+
+#### DPC++ Headers
+
+The same DPC++ Headers are used for native and emulated specialization constants
+and their design is decribed in the corresponding sub-section of "Support for
+native specialization constants" section.
+
+However, that part of the document doesn't describe the third argument of
+`__sycl_get*2020SpecConstantValue` intrinsics: it is a pointer to a runtime
+buffer, which holds values of all specialization constants and should be used
+to retrieve their values in device code.
+
+This pointer is stored within `kernel_handler` object and it is initialized only
+if our target doesn't support native specialization constants.
+Since `kernel_handler` object is not captured by SYCL kernel funtion, it means
+that we are not able to employ some header-only solution here and need help of
+the compiler.
+
+DPC++ FE searches for functions marked with `sycl_kernel` attribute to handle
+them and turn into entry points of device code.
+
+
+#### DPC++ FE
+
+When we compile code for target which doesn't support native specialization
+constants, DPC++ FE should look for `kernel_handler` argument in functions
+marked as `sycl_kernel`. If such argument is present, it means that this kernel
+can access specialization constants and therefore we need to:
+- generate one more kernel argument for passing a buffer with specialization
+  constants values.
+- create `kernel_handler` object
+  
+  **TODO**: this item should be done for native specialization constants as
```

Comment:
Please clearly specify the requirements for the 2 cases here i.e. requirements for native support vs not native support. From my understanding the additional argument (and corresponding __init call) is only generated when there is no native support for specialization constants, and the default constructor is used to initialize the kernel_handler object in openCL kernel when there is native support. Please confirm/elaborate on this in the spec.

Reply:
Do not see a reason why we would need to have additional argument and __init call for targets that support specialization constants natively. Maybe I'm missing something, looking into it.

Reply:
Tried to clarify in the doc. Please, take a look.